### PR TITLE
feat: align the rest of the ContentTypeParser methods with `.add`

### PR DESCRIPTION
**Pull request by gurgunday**
Per @climba03003's comment, https://github.com/fastify/fastify/pull/4477#issuecomment-1972791560

[We now trim and lowercase the values of _user-added_ ContentTypeParsers](https://github.com/fastify/fastify/pull/5329) so that weird scenarios like [this](https://github.com/fastify/fastify/pull/4477#issuecomment-1378545312) are not possible

The same behavior should be exhibited by `.remove` and `.hasParser` for consistency

After the merge, follow-up PRs:

- [ ] Fix a bug in contentType parser matching — https://github.com/fastify/fastify/pull/5371
- [ ] Throw if the passed argument type is not a string or regexp in `hasParser` like the other methods — https://github.com/fastify/fastify/pull/5372

### CODE REVIEW

**Comment by gurgunday** [comment]
I can revert https://github.com/fastify/fastify/pull/5345/files#diff-40e49c52c74253254f63de155bd3c7183d0d81936d8f272cfd6775b6af65bdb5R210 but it is literally semantically equivalent

And https://github.com/fastify/fastify/pull/5345/files#diff-40e49c52c74253254f63de155bd3c7183d0d81936d8f272cfd6775b6af65bdb5R187 doesn't do string comparison, so the performance is [equivalent](https://perf.link/#eyJpZCI6IjR6dXYxdnZ1aDY3IiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGRhdGEgPSBbLi4uQXJyYXkoMTAwMCkua2V5cygpXSIsInRlc3RzIjpbeyJuYW1lIjoiMSIsImNvZGUiOiJkYXRhICYmIHR5cGVvZiBkYXRhLnRoZW4gPT09IFwiZnVuY3Rpb25cIiIsInJ1bnMiOls0MDE5MDAwLDc5MTAwMCw2NjY4MDAwLDUwNzcwMDAsMTQzMDAwMCw2Nzc0MDAwLDczNjMwMDAsMTA2NTIwMDAsNjk2NTAwMCw3MDQ5MDAwLDU4ODAwMDAsMTI1ODAwMCw3ODAzMDAwLDMwMDkwMDAsODQ4MDAwMCwzNTUwMDAwLDc3MDMwMDAsODY5MDAwMCwyNTMzMDAwLDI2MTkwMDAsMTU3NDgwMDAsMTE1MTAwMDAsMTYyNjYwMDAsMTQ3NDUwMDAsMTczMTEwMDAsMTQ4MTAwMCwxMDAwODAwMCwxMzAzNzAwMCwxMzE3NjAwMCw4MzA5MDAwLDExNzY5MDAwLDE3OTcwMDAwLDE0MjIxMDAwLDEyNzU3MDAwLDY4NjkwMDAsMTMwMzIwMDAsMzAyMTAwMCwxMDEyMjAwMCw3ODQzMDAwLDExMDUwMDAwLDExNTg0MDAwLDE0MzAwMDAsNzAyMDAwMCwxMjAxNTAwMCwxMzk1OTAwMCwxMDkzMDAwMCwxMzgzODAwMCw2NDczMDAwLDgzOTcwMDAsOTM1OTAwMCw4Njg5MDAwLDEwNTAzMDAwLDc1MjEwMDAsMzAyMTAwMCwxMzUxMDAwMCwxMjQ0NzAwMCwxMzI1NDAwMCwxMjIzODAwMCwyNjM0MDAwLDExMTc5MDAwLDEyNjYzMDAwLDEzMjg1MDAwLDE1NDA4MDAwLDEwNzUwMDAwLDIwMTg3MDAwLDEyNTA1MDAwLDExNDMyMDAwLDkxMjQwMDAsMTA1OTcwMDAsMTQ3NzAwMCwxNjc0NjAwMCwxNjMwNjAwMCwxNTcyMTAwMCwxMDUwOTAwMCwxODgwMDAwLDk5NjgwMDAsMTA5MDkwMDAsOTU3NTAwMCw1NTk3MDAwLDExMTc3MDAwLDUzNDIwMDAsMTU5MzYwMDAsMTI3NTgwMDAsMTQ0MzMwMDAsMTM3ODkwMDAsOTA3MzAwMCwxOTQ4MDAwLDE2NzIwMDAwLDE2MzI5MDAwLDEyMTIwMDAwLDE2MzA1MDAwLDE0MzAwMDAsMTcxMzgwMDAsMTE2NzEwMDAsMTEzNjkwMDAsMTIzMTMwMDAsMTAzMjMwMDAsMTg2NjMwMDAsMTAyMDMwMDAsMTQwMTQwMDBdLCJvcHMiOjk5NjI1MjB9LHsibmFtZSI6IjIiLCJjb2RlIjoidHlwZW9mIGRhdGE%2FLnRoZW4gPT09IFwiZnVuY3Rpb25cIiIsInJ1bnMiOls0MzgzMDAwLDQyODAwMCw1NDAxMDAwLDU1NzgwMDAsMTQ0ODAwMCw2Mzk4MDAwLDc4MjcwMDAsMTA3NzAwMDAsNjgzMzAwMCw4ODc0MDAwLDY1NDcwMDAsMTM3NjAwMCw3OTIyMDAwLDMwMTcwMDAsOTMwNTAwMCwzOTIzMDAwLDc4MDIwMDAsOTEzMDAwMCwyNjA0MDAwLDI2OTgwMDAsMTU3NDkwMDAsMTE4ODEwMDAsMTYzNjEwMDAsMTQxNzYwMDAsMTc1ODMwMDAsMTUxNzAwMCwxMDM0MjAwMCwxMzU0NDAwMCwxMzI0MzAwMCw4MjgzMDAwLDEwNjU5MDAwLDE3OTk3MDAwLDE0MjU5MDAwLDEyNTM1MDAwLDcyMDMwMDAsMTMxMjQwMDAsMzA4NTAwMCwxMDczODAwMCw4NDgwMDAwLDExMjQ0MDAwLDEyMTM5MDAwLDE0NDgwMDAsNzExNzAwMCwxMjA3NDAwMCwxNDMzOTAwMCwxMDk1NTAwMCwxNDAzOTAwMCw2MDM0MDAwLDgwMjgwMDAsOTUwOTAwMCw4NjgwMDAwLDEwNjc4MDAwLDEwMDU3MDAwLDMxNzYwMDAsMTQxMjkwMDAsMTI3ODkwMDAsMTMzNzkwMDAsMTIwOTIwMDAsMjU1MjAwMCwxMDk1MzAwMCwxMjk4MjAwMCwxMjYyOTAwMCwxNTU0MzAwMCwxMTEzNzAwMCwyMDI5MTAwMCwxMjk1MjAwMCwxMDc3MDAwMCw5MTg3MDAwLDEwNzMzMDAwLDE1OTQwMDAsMTY3MTQwMDAsMTYxMTcwMDAsMTUyMTEwMDAsMTAzMTEwMDAsMjU3MTAwMCwxMTcwOTAwMCwxMjE1NjAwMCw5NzAzMDAwLDU3OTMwMDAsMTA3ODUwMDAsNTQ4NDAwMCwxNjMzMDAwMCwxMzg3ODAwMCwxNDM3NTAwMCwxMzUxNzAwMCwxMDcwODAwMCwyMDAxMDAwLDE2NzcwMDAwLDE2ODczMDAwLDEyNjc1MDAwLDE1Mzc0MDAwLDE0NDgwMDAsMTY0NDYwMDAsMTE3NDEwMDAsMTEwOTMwMDAsMTIyMDcwMDAsODEyODAwMCwxODg5NjAwMCwxMDA5MzAwMCwxMzc4NjAwMF0sIm9wcyI6MTAwOTE0NTB9XSwidXBkYXRlZCI6IjIwMjQtMDMtMDJUMTE6NTM6MjIuOTUwWiJ9)

File: lib/contentTypeParser.js
Comment (mcollina) [inline lib/contentTypeParser.js:98]: i feel this will slow us down

Comment (climba03003) [reply lib/contentTypeParser.js:98]: From my point of view, `.trim` should place way lower place and it should be another `string`.
So, it would cache the original header and utilize the cache.

**Comment by gurgunday** [comment]
I checked it and the Content-Type is matched correctly even if there are surrounding spaces - so trimming happens somewhere

`getParser` is only called internally, as a result, I removed any modifications from it

**Comment by gurgunday** [comment]
Ok everything is much better now, and faster than `next` probably - note that getParser is internal and only takes a string contentType

Please take a look

**Comment by gurgunday** [comment]
I will change the order of the methods for better readability after getting an approval, so please don't merge 🙏

**Comment by gurgunday** [comment]
Benchmark isn't working on `next` I believe

File: lib/contentTypeParser.js
Comment (gurgunday) [reply lib/contentTypeParser.js]: I checked it and the Content-Type is matched correctly even if there are surrounding spaces - so trimming already happens somewhere

getParser is only called internally, as a result, I removed any behavioral changes from it, focusing on the public API

File: lib/contentTypeParser.js
Comment (Uzlopak) [inline lib/contentTypeParser.js:211]: please explain

Comment (gurgunday) [reply lib/contentTypeParser.js:211]: `Number(undefined) === NaN`

File: lib/contentTypeParser.js
Comment (gurgunday) [reply lib/contentTypeParser.js:188]: It's not slower either, it just reads better imo, do you disagree?

There is no string comparison as I mentioned previously

File: lib/contentTypeParser.js
Comment (metcoder95) [inline lib/contentTypeParser.js:142]: Maybe avoid the allocation if used only for the `if` statement?

**Review by metcoder95 — COMMENTED** [review]
LGTM, but I'd prefer to wait for @climba03003's review

**Comment by climba03003** [comment]
I need to test it when I have time.
Just want to figure out where did the `.trim` happen before `.getParser`.

**Comment by gurgunday** [comment]
> Just want to figure out where did the .trim happen before .getParser.

Please do let me know when you find it 😁

**Comment by gurgunday** [comment]
@climba03003 I think node does it since this returns the trimmed version:

```js
  fastify.post('/', (req, reply) => {
    console.log(req.raw.header['content-type'])
    reply.send(req.body)
  })
```

**Comment by gurgunday** [comment]
I've found a bug with the `contentType` matching, so it would be appreciated if we can get this merged

I will be concentrating on v5 for a few days

**Review by climba03003 — APPROVED** [review]

**Review by metcoder95 — APPROVED** [review]

**Comment by gurgunday** [comment]
Merged at https://github.com/fastify/fastify/pull/5372

**Pull request closed without merging**
